### ISSUE TITLE
Using Get-Command in script yields unwanted / irrelevant suggestions at prompt after script has run

### ISSUE THREAD

**Issue by sba923** [PowerShell/PowerShell#19862]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

After running a script (that does a number of `Get-Command` invocations, probably related), I get an unexpected (unwanted, irrelevant...) suggestion.

Here's the script:

```powershell
Get-ChildItem -LiteralPath $PSScriptRoot -File -Filter '*.ps1' | Where-Object {
    $sbversion = Get-Command $_.Name -ErrorAction SilentlyContinue
    if ($null -ne $sbversion)
    {
        $sbversion_i = Get-Item -LiteralPath $sbversion.Path
        if ($sbversion_i.LastWriteTime -gt $_.LastWriteTime)
        {
            Copy-Item -Verbose -LiteralPath $sbversion.Path -Destination $_.FullName
        }
    }
}
```

(The script updates all scripts found in the folder containing the script if there are more recent versions of the same scripts on the `PATH`)

### Expected behavior

```console
No suggestion should be displayed
```

### Actual behavior

```console
Suggestion [3,General]: The command zr_4_int.ps1 was not found, but does exist in the current location. PowerShell does not load commands from the current location by default. If you trust this command, instead type: ".\zr_4_int.ps1". See "get-help about_Command_Precedence" for more details.
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.4
PSEdition                      Core
GitCommitId                    7.3.4
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![image](https://github.com/PowerShell/PowerShell/assets/12860484/9c55a09f-572b-4092-acd5-b687429ef1f4)

**Comment by 237dmitry** [comment]
Try to change `$_.Name` to `$_.FullName`

**Comment by sba923** [comment]
This would totally break the logic of the script, whose purpose is to look for scripts on the `PATH` with the same names as the ones in the (current) folder.

This implements the desired behavior, and triggers the unwanted suggestion:

```powershell
Get-ChildItem -LiteralPath $PSScriptRoot -File -Filter '*.ps1' | Where-Object {
    $sbversion = Get-Command $_.Name -ErrorAction SilentlyContinue
    if ($null -ne $sbversion)
    {
        Write-Host -ForegroundColor Green ("Found '{0}' on PATH as '{1}'" -f $_.Name, $sbversion.Path)
    }
    else {
        Write-Host -ForegroundColor Red ("No '{0}' found on PATH" -f $_.Name)
    }
}
```

![image](https://github.com/PowerShell/PowerShell/assets/12860484/39737838-db35-4f4f-a304-556ca127dee6)


This doesn't trigger the unwanted suggestion, but doesn't implement the desired logic:

```powershell
Get-ChildItem -LiteralPath $PSScriptRoot -File -Filter '*.ps1' | Where-Object {
    $sbversion = Get-Command $_.FullName -ErrorAction SilentlyContinue
    if ($null -ne $sbversion)
    {
        Write-Host -ForegroundColor Green ("Found '{0}' on PATH as '{1}'" -f $_.Name, $sbversion.Path)
    }
    else {
        Write-Host -ForegroundColor Red ("No '{0}' found on PATH" -f $_.Name)
    }
}
```

![image](https://github.com/PowerShell/PowerShell/assets/12860484/db6a5b54-67bd-485f-b9fc-9779d996f02c)

Strangely enough, this does *not* trigger a suggestion:

```powershell
@('xyz.ps1', 'df.ps1') | Where-Object {
    $sbversion = Get-Command $_ -ErrorAction SilentlyContinue
    if ($null -ne $sbversion)
    {
        Write-Host -ForegroundColor Green ("Found '{0}' on PATH as '{1}'" -f $_, $sbversion.Path)
    }
    else {
        Write-Host -ForegroundColor Red ("No '{0}' found on PATH" -f $_)
    }
}
```

![image](https://github.com/PowerShell/PowerShell/assets/12860484/ed566319-1c22-4219-ae14-b08c74281029)

**Comment by 237dmitry** [comment]
You can test files with `Get-FileHash`

```powershell
pushd BackupDir

Get-Childitem *.ps1 | Foreach-Object {                                     

  $bin = (Get-FileHash ($f = Get-Command $_.Name).Path).Hash
  $arc = (Get-FileHash $_).Hash
  if ($bin -ne $arc) { $f.Path }
  }
```

![ss-20230627151237](https://github.com/PowerShell/PowerShell/assets/78153320/b7f5e836-fe55-4960-8d6e-cb15de8f2823)

**Comment by sba923** [comment]
> You can test files with `Get-FileHash`
> 
> ```powershell
> pushd BackupDir
> 
> Get-Childitem *.ps1 | Foreach-Object {                                     
> 
>   $bin = (Get-FileHash ($f = Get-Command $_.Name).Path).Hash
>   $arc = (Get-FileHash $_).Hash
>   if ($bin -ne $arc) { $f.Path }
>   }
> ```
> 
> ![ss-20230627151237](https://user-images.githubusercontent.com/78153320/249146478-b7f5e836-fe55-4960-8d6e-cb15de8f2823.png)

This is getting totally off-topic.

This issue is about the suggestion engine's behavior.

**Comment by 237dmitry** [comment]
> This issue is about the suggestion engine's behavior.

Do you see the suggestion in the screenshot? It also has `Get-Command $_.Name`.

**Comment by 237dmitry** [comment]
Remove this suggestions...

```powershell
PS > 'fake', 'not exists', 'bash' | % { if ($c = gcm $_ -ea 0) { $c.Path }}
/usr/bin/bash

 PS > foreach ($command in 'fake', 'not exists', 'bash')  { if ($c = gcm $command -ea 0) { $c.Path }}
/usr/bin/bash
```

**Comment by SteveL-MSFT** [comment]
The new FeedbackProvider added in 7.4 replaces the legacy suggestion system which resolves this

**Comment by 237dmitry** [comment]
In 7.3.* you can disable `PSCommandNotFoundSuggestion`, but this will require a restart of pwsh.

```Disable-ExperimentalFeature PSCommandNotFoundSuggestion```

**Comment by sba923** [comment]
> In 7.3.* you can disable `PSCommandNotFoundSuggestion`, but this will require a restart of pwsh.
> 
> `Disable-ExperimentalFeature PSCommandNotFoundSuggestion`

This doesn't seem to work:

![image](https://github.com/PowerShell/PowerShell/assets/12860484/06b4bf43-7294-4440-9baa-de376a50070d)

**Comment by 237dmitry** [comment]
> This doesn't seem to work:

Weird, it doesn't really work. I've tried before and it worked interactively (not from a script). Otherwise, I wouldn't recommend it.

strange behavior:

![Screenshot 2023-06-28 123809](https://github.com/PowerShell/PowerShell/assets/78153320/02bc771c-56e0-4e40-b4cf-a32a6174d4d6)

But after executing your script (without copying) suggestions returned:

![Screenshot 2023-06-28 125002](https://github.com/PowerShell/PowerShell/assets/78153320/df895024-a91e-4a74-9d97-3fc6de337eee)

**Comment by sba923** [comment]
> Does it break logic:
> `$sbversion = Get-Command "$($_.Name)*" -ErrorAction SilentlyContinue`

In my _very limited_ test setup, the behavior is the same, but I'm kinda wondering what the trailing `*` could change...

![image](https://github.com/PowerShell/PowerShell/assets/12860484/79b04850-e11a-4948-b7d7-3f57e07909eb)

**Comment by 237dmitry** [comment]
> I'm kinda wondering what the trailing * could change...

This is not exact search, rather like-search and can find something like file.ps1xml

PS. For previous post, Get-Command outputs suggestion if file in current directory. This was my wrong advice.
```
gcm fake -ea 0    # no suggestion if in current location there is not fake.ps1 
```

**Comment by sba923** [comment]
> This is not exact search, rather like-search and can find something like file.ps1xml

`Get-Command` does *not* look for / find `.ps1xml` files!

![image](https://github.com/PowerShell/PowerShell/assets/12860484/e9284d0a-0e20-4f21-a4cc-d58a6fb7893c)

**Comment by sba923** [comment]
Did I say I was running 7.3.4 on Windows?

**Comment by 237dmitry** [comment]
> Did I say I was running 7.3.4 on Windows?

Ok. Situation on Windows:

![Screenshot 2023-06-28 173228](https://github.com/PowerShell/PowerShell/assets/78153320/97ec9d02-1bf8-489c-b77a-d90e04d92f88)

**Comment by 237dmitry** [comment]
> I don't understand why the unwanted suggestion shows up without the * and disappears with the *

I think the cmdlet is not looking for the exact result, but a similar one. Therefore, there are no errors and suggestions.

**Comment by sba923** [comment]
> I think the cmdlet is not looking for the exact result, but a similar one. Therefore, there are no errors and suggestions.

So a failure to find a command for a _wildcard_ match returns `$null` _without_ an error, a failure for an _exact_ match returns $null` _with_ an error?

**Comment by 237dmitry** [comment]
> So a failure to find a command for a wildcard match returns $null without an error, a failure for an exact match returns $null` with an error?

Yes. But `*` returns array and failure exact returns `$null`

**Comment by sba923** [comment]
> Yes. But `*` returns array and failure `exact match` returns `$null`

That's not what I see here:

![image](https://github.com/PowerShell/PowerShell/assets/12860484/bc741866-6dd3-4c19-9a67-fff0a10d394b)

**Comment by 237dmitry** [comment]
> That's not what I see here

I'm sorry, this is all my inattention. I mixed up local and remote consoles.

![Screenshot 2023-06-28 223824](https://github.com/PowerShell/PowerShell/assets/78153320/8989af98-853c-4641-b3e7-b89b3b4582b7)

**Comment by sba923** [comment]
> The new FeedbackProvider added in 7.4 replaces the legacy suggestion system which resolves this

I confirm the issue's not present in 7.4.0-preview.4:

![image](https://github.com/PowerShell/PowerShell/assets/12860484/077e5830-31c0-45d3-b04e-268585c507e3)

**Comment by sba923** [comment]
> So this is now mainstream and not experimental. How do I turn it off? It's so annoying.

@cow1337killer3 This issue is closed (since it had been fixed in 7.4.0). I suggest you hop on to https://github.com/PowerShell/PowerShell/issues/24712.
